New roles

pwalczysko opened the pull request:
# What this PR does

This PR introduces the new roles for facility manager workflows. These workflows are facilitated by creating non-full-admin users (called internatlly light admins) which then can execute tasks like ImportAs (importing for other users) or organizing data of other users or users into groups in any group they like. It goes about the permissions restrictions by having a set of privileges in experimenter_config table in the DB. For each of those privileges (or restrictions, they can be either true or false), there is an expected behaviour. 

The accompanying tests deal with two problematics:

  - unwanted expansion of permissions (e.g. by sudoing as true admin, the light admin will NOT get full admin privileges), these are covered mostly in @mtbc 's tests
  - desired workflows feasibility (e.g. can I, equipped with privileges Sudo, Chown, WriteFile and WriteOwned go through a meaningful workflow of ImporterAs ?), these are covered in my tests

# Testing this PR

Mainly check that the accompanying tests as well as all the other integration tests are passing and make sense.

# Related reading

[design issue](https://github.com/openmicroscopy/design/issues/62)
[trello card](https://trello.com/c/t0nT7KYa/133-new-role)
The .ppt explanation of the tests is at ``/ome/team/pwalczysko/new-roles-tests-explanation/Explanation-Of-TestsNewRole.ppt``
Edit: also the Power Point presentation is now as a [g.doc](https://docs.google.com/presentation/d/18kc75LHiItkHr_dvG1R6jW1qmTNayan8x4ZQC6s7XmI/edit#slide=id.g1b11290c7c_2_96) better accessible.
See here for the [table](https://docs.google.com/spreadsheets/d/1HadUCZlIrJoetDZrWfEFYrralCakqLGWPpjK_hHjaqM/edit#gid=415364306) of features.

1. background for understanding this PR
The facility managers using omero, as well as persons such as image analysts need a free access to other user's data, and in some degrees freedom to create containers, change ownership of the data etc., without having the full powers and responsibilities of the full administrator of OMERO.

cc @jburel @joshmoore @mtbc @bramalingam 

Not ready for review yet. Will add some .ppt clarifying the tests graphically. Also, @mtbc is still working on new features.

--breaking

mtbc committed:
move SQL scripts ready for DB patch 14

mtbc committed:
prepare SQL scripts for patch 14

also simplify end-script message-printing

mtbc committed:
add light admin privileges and session sudoers to OMERO model

mtbc committed:
generate new database schema for patch 14

mtbc committed:
add SQL for actual upgrade DDL

mtbc committed:
add initial light admin privileges to OMERO model

mtbc committed:
ease debugging for server-side use of enumerations

includes enumeration string value in toString()

mtbc committed:
extend the IAdmin API for light administrator privileges

mtbc committed:
add a light administrator privileges helper utility bean

mtbc committed:
allow using admin privs, other enums, in hash-based collections

mtbc committed:
implement new IAdmin methods for light admin privileges

mtbc committed:
graph traversal cannot ignore permissions for light administrators

mtbc committed:
have server session management preserve sudoer information

mtbc committed:
add ModifyUser privilege checking to IAdmin implementation

mtbc committed:
build Sudo privilege checking into session bean

mtbc committed:
build ModifyUser, WriteOwned checks into security system

mtbc committed:
fix remaining unit tests after constructor changes

mtbc committed:
add tests for working with light admin privileges via IAdmin

mtbc committed:
add testing for ModifyUser, Sudo, WriteOwned privileges

mtbc committed:
decrease log level for unknown privilege checks

mtbc committed:
thread admin privileges helper into graph requests helper

mtbc committed:
help for graphs requests with checking admin privileges

mtbc committed:
make set from LightAdminPrivileges.getSessionPrivileges immutable

mtbc committed:
make light admin privilege test code slightly clearer

mtbc committed:
add light admin privileges for Chgrp and Chown

mtbc committed:
check for Chgrp and Chown privilege in OMERO interceptor

mtbc committed:
check for Chgrp and Chown privilege in graphs requests

mtbc committed:
add integration testing for Chgrp and Chown admin privileges

mtbc committed:
fix Javadoc for WriteOwned privilege tests

mtbc committed:
simplify how tests add user to system group

mtbc committed:
move Chgrp, Chown privilege tests to list privileges in order

mtbc committed:
add initial testing for WriteFile light admin privilege

mtbc committed:
have model mock factory suffix file paths with usual separator

mtbc committed:
add WriteFile checking to basic security system

mtbc committed:
make variable final in test

mtbc committed:
remove unnecessary method annotation

mtbc committed:
add test for robustness of script service listing scripts

mtbc committed:
prevent creating objects in inaccessible groups

mtbc committed:
add light admin privilege checking to script service

mtbc committed:
extend integration tests to cover various WriteFile cases

mtbc committed:
extract integration test importFileset methods into superclass

mtbc committed:
add light admin WriteFile test for changing file checksum

mtbc committed:
fix error message to not mention null

mtbc committed:
internal service factory creation should follow group name

mtbc committed:
have import request accept and propagate group context

mtbc committed:
have import process propagate call context to import request

mtbc committed:
minor cleanup of light administrator integration tests

mtbc committed:
add WriteFile restriction test for FS import

mtbc committed:
fix reader name in import process for integration testing

mtbc committed:
add integration tests for the ReadSession privilege

pwalczysko committed:
Add test for ImporterAs workflow, adjust parent class

pwalczysko committed:
Delete unused imports, change author, delete unused methods

pwalczysko committed:
Add tests for ImporterAs ability to link and delete data with Sudo only

pwalczysko committed:
Add tests for edit name and chgrp for importerAs Sudo only

pwalczysko committed:
Add test for additional chgrp privilege on chgrp behaviour

mtbc committed:
allow light admin privilege check without caching session

mtbc committed:
build sudo-aware session load filter into basic ACL voter

mtbc committed:
fix minor typo in comment

mtbc committed:
build light admin security filter into basic security system

Rather than including in security filter holder, which chooses
among alternative filters, this light admin security filter
is added to a /collection/ of filters used by the basic security
system.

mtbc committed:
rework session querying to be sudo- and ReadSession-aware

mtbc committed:
rework light admin testing

- tidy abstract server test roles management
- tidy handling of sudo in light admin tests

mtbc committed:
add integration tests for the ReadSession privilege

mtbc committed:
wrap longest lines in integration test code (whitespace only)

mtbc committed:
switch internal service factory to groups by ID, not name

avoids risky session factory destruction

pwalczysko committed:
Add test for ImporterAs using Chown permission after import

mtbc committed:
add IAdmin.getCurrentAdminPrivileges() with integration test

mtbc committed:
test WriteFile privilege enforcement via newTable

mtbc committed:
intercalating underscores as single character

mtbc committed:
have BasicACLVoter.allowChmod respect modify/write privileges

mtbc committed:
adjust Chgrp2I to take better account of Chgrp privilege

pwalczysko committed:
Simplify the chown test, add similar logic to chgrp test, catch server bug

pwalczysko committed:
Merge branch 'roles' of github.com:mtbc/openmicroscopy into new-role

pwalczysko committed:
Make the tests work with recent Marks changes in parent clesses

pwalczysko committed:
Adjust chgrp test for ImporterAs and aligh it with chown test

pwalczysko committed:
Delete print statements, correct comments

mtbc committed:
improve links from Javadoc

mtbc committed:
add IAdmin.createLightSystemUser method with test

mtbc committed:
better error checking/reporting for mistyped privileges

pwalczysko committed:
Add tests for ImporterAs workflows with No Sudo

pwalczysko committed:
Add tests for creation and targeting of Dataset to NoSudo ImporterAs workflows

pwalczysko committed:
Add test for Delete behavior of a sudoing and non-sudoing light admin

pwalczysko committed:
Add a supplementary test for light admin linking users objects

pwalczysko committed:
Hieararchy chown failing for light admin, passing for full admin, also psssing if light admin owns hierarchy

pwalczysko committed:
Permutate over all group types in combined privilege tests

mtbc committed:
for chgrp,chown make screen-plate policy match other containers

mtbc committed:
to chgrp,chown permissions checks add light admin workaround

mtbc committed:
protect interceptor from being passed an unloaded event

mtbc committed:
protect basic ACL voter from being passed an unloaded event

mtbc committed:
fix mistake in test method Javadoc

mtbc committed:
the switch to proper FakeReader reduces retained client paths

mtbc committed:
correct accidentally duplicated constructor argument

mtbc committed:
use top-level OMERO binary repository in WriteFile testing

mtbc committed:
simplify setup noting all light admin privilege

pwalczysko committed:
Merge branch 'roles' of github.com:mtbc/openmicroscopy into new-role

mtbc committed:
have Javadoc reflect change in method

pwalczysko committed:
Adjust tests to test Mark's fix of the chown hierarchy bugs

mtbc committed:
Merge remote-tracking branch 'remotes/origin/develop' into roles

Conflicts:
	components/blitz/src/omero/cmd/graphs/Chown2I.java
	components/model/resources/mappings/meta.ome.xml
	components/tools/OmeroJava/test/integration/AbstractServerTest.java

pwalczysko committed:
Merge branch 'roles' of github.com:mtbc/openmicroscopy into new-role

pwalczysko committed:
Fix the testImporterAsNoSudoLinkInTargetGroup : the full hierarchy chown does not work in this case

pwalczysko committed:
Rename test accordingly to its functions

pwalczysko committed:
Fix the comment in the test according to test function

mtbc committed:
make security system aware of script repository UUID

pwalczysko committed:
Widen Edit workflow test for NoSudo, permutate over all groups and permission combinations

pwalczysko committed:
Fix the lightAdminEdit test for ownership of the Project

pwalczysko committed:
Go over all group permissions in the testImporterAsSudoCreateImport

pwalczysko committed:
Adjust the testDataOrganizerChownAll to truly chown all belonging to a user

jburel commented:
Few unit test failures (travis) e.g.
``
FAILED: testAddLog on null(ome.server.utests.sec.SecuritySystemTest)
[nomemorytestng] org.jmock.core.DynamicMockError: mockLocalQuery: no match found
[nomemorytestng] Invoked: ome.api.IQuery.findByQuery(<FROM Session s LEFT OUTER JOIN FETCH s.sudoer WHERE s.id = :id>, <PARAMS:id=1 >)
[nomemorytestng] Allowed:
``

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -0,0 +1,1305 @@
+         * the import into the just created Dataset.
+         * Check thus that the light admin can import and write the original file
+         * on behalf of the normalUser and into the group of normalUser */
+        client.getImplicitContext().put("omero.group", Long.toString(normalUser.groupId));
```
Didn't we already do this above?

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -0,0 +1,1305 @@
+        lightAdmin = loginNewAdmin(isAdmin, permissions);
+        /* set up the project as the normalUser */
+        loginUser(normalUser);
+        client.getImplicitContext().put("omero.group", Long.toString(normalUser.groupId));
```
Isn't the user *already* working in this group without needing this set?

mtbc commented:
@jburel: Gah, I hate these weird tests of implementation details instead of actual behavior. @pwalczysko: Don't worry about those failures, they're just something tedious I'll have to get around to looking at. In the meantime maybe we can set breaking to ignore travis failures.

mtbc committed:
map OriginalFile.repo from DB into OMERO object model

mtbc committed:
have integration tests use uniquely named files

Allows tests to run multiple times without running foul of
name uniqueness constraints.

mtbc committed:
fix original file creation for newly mapped "repo" property

Threads Hibernate session into Repository DAO file registration.

Conservatively continuing to set file repo via SQL for the interim.
Must therefore merge that change into Hibernate object state.

pwalczysko commented:
> In the meantime maybe we can set breaking to ignore travis failures.

Done courtesy of @sbesson - Thank you.

mtbc committed:
prevent HQL setting of OriginalFile.repo property

Before allowing write access to OriginalFile.repo we
must first think through the permissible conditions.

pwalczysko committed:
Remove redundant group context specifications cf. Mark's comments

pwalczysko committed:
Adjust data provider description

pwalczysko committed:
Make chown passing test more explicit in testDataOrganizerChownAll

pwalczysko committed:
Add ownership checks for original file in testImporterAsSudoChown, add explanation to testDataOrganizerChownAll

mtbc committed:
use unique object names in tests

mtbc committed:
add new WriteScriptRepo light admin privilege

mtbc committed:
add permissions tests for WriteScriptRepo privilege

largely mirroring WriteFile tests which apply to other repos

mtbc committed:
expand server permissions checking for WriteScriptRepo

pwalczysko committed:
Improve explanatory comments in testImporterDelete

mtbc committed:
split Delete* light admin privileges out  from Write* ones

mtbc committed:
adjust Basic ACL Voter to separate Write* from Delete*

mtbc committed:
adjust integration tests for Write* and Delete* privileges

mtbc committed:
reorder integration tests within class

pwalczysko committed:
Make the testImporterAsSudoChgrp more concise, improve comments

pwalczysko commented:
The .ppt explanation of the tests is at ``/ome/team/pwalczysko/new-roles-tests-explanation/Explanation-Of-TestsNewRole.ppt`` (still working on it)

pwalczysko committed:
Check also for original file group in testImporterAsSudoChgrp

mtbc committed:
use script repository (easily identified) in repo UUID test

mtbc committed:
more correctly identify the script repository among its kin

mtbc committed:
improve how tests find the desired kind of repository

mtbc commented:
@pwalczysko: Happy for you to move this to a different branch corresponding with devspace @aleksandra-tarkowska is kindly setting up for us. (We can tweak branch in its CI config.)

mtbc committed:
re-permit users writing to world-writable groups

mtbc committed:
be clearer about expecting to find repositories

mtbc committed:
adjust "fs sets" test to expect subsets of IDs

can't use straight equality because report will also list
filesets that are in publicly readable groups

mtbc committed:
fix minor typo in comment

pwalczysko committed:
Add tests for no sudo workflow to Create part of testImporterAsSudoCreateImport

mtbc committed:
fix security system unit tests with dummy sessions

joshmoore commented on components/blitz/src/omero/cmd/graphs/GraphHelper.java:
```diff
@@ -51,15 +56,34 @@
+     */
+    public boolean checkIsAdministrator(AdminPrivilege requiredPrivilege) {
+        final EventContext eventContext = helper.getEventContext();
+        if (!eventContext.isCurrentUserAdmin()) {
```
So any privilege makes a user an admin, or for any privilege to be considered valid, the user must be an admin? To some extent, I wonder if the `ec.isCurrentUserAdmin` method doesn't need to be deprecated in favor of this method (refactored up) since the rest of the code-base will be assuming a binary state as opposed to the new spectrum.

mtbc replied:
Yes, one thing I got from the "restrictions" discussion was that light admins are members of `system`. I did experiment with trying to introduce a string set of privileges in event context but got confused over interactions with sudo and intimidated by fear of getting things wrong. My current thinking is to try to make new roles minimally invasive for 5.3 then once we have more time and experience to fit them in more naturally for 5.4 and your suggestion is a prime candidate for how that needs to be done.

mtbc commented:
While @pwalczysko's away I might be able to fix test failures via #5054.

joshmoore commented on components/server/src/ome/logic/AdminImpl.java:
```diff
@@ -634,13 +640,34 @@ public long createSystemUser(Experimenter newSystemUser) {
+        restrictions.removeAll(privileges);
+        final List<NamedValue> userConfig = new ArrayList<NamedValue>(restrictions.size());
+        for (final AdminPrivilege restriction : restrictions) {
+            userConfig.add(new NamedValue(restriction.getValue(), Boolean.toString(false)));
```
What does this currently look like in the DB? I'd imagine a prefix might be useful for long-term upgrades, etc.

mtbc replied:
You're thinking the prefix to distinguish from other kinds of user configuration? Maybe even no prefix in `AdminPrivilege.value` but only in `Experimenter.config.name`? Could look into that.

joshmoore commented on components/server/src/ome/logic/AdminImpl.java:
```diff
@@ -1442,4 +1557,29 @@ private void adminOrPiOfNonUserGroups(ExperimenterGroup defaultGroup,
+        if (isAdmin()) {
+            final String hql = "FROM Session s LEFT OUTER JOIN FETCH s.sudoer WHERE s.id = :id";
+            final Parameters params = new Parameters().addId(getEventContext().getCurrentSessionId());
+            final ome.model.meta.Session session = iQuery.findByQuery(hql, params);
```
Has there been an performance impact by querying for the privileges? If so, we should likely re-use the `SessionCache` infrastructure to pre-load them.

mtbc replied:
Not that I've noticed but I haven't much checked (can changes in CI timings help?) and the internal security infrastructure, which I figure is where the frequent checks occur, is making use of an explicit cache of them, and where they're not in an OMERO cache I'd guess they're probably soon in a Hibernate one. Could certainly look at preloading sudoer somehow? Maybe even preloading privileges depending on how session cache stuff would handle changes in privileges during long-lived session.

joshmoore commented:
Few design-level comments made. I haven't yet reviewed the server-side tests.

mtbc committed:
fix logic in AdminImpl user permissions test

mtbc commented:
`LightAdminRolesTest` alone takes over an hour to run. Might need to reduce parameter combinations for tests that use expensive operations (e.g., import image, upload script).

mtbc committed:
to test script upload avoid working on another test's script

mtbc commented:
@sbesson: If we could now switch "breaking" back to requiring travis passes then this PR will happen to come back in to the merge once @pwalczysko has merged in my latest commits and adjusted for the `Write*`/`Delete*` privilege split.

sbesson commented:
@mtbc changes done

mtbc committed:
limit system type permissions check skipping to job types only

mtbc committed:
add ModifyGroup, ModifyGroupMembership light admin privileges

mtbc committed:
add tests for new user management light admin privileges

mtbc committed:
add core security system checks for user management privileges

mtbc committed:
add admin service checks for user management privileges

mtbc committed:
reorder integration tests

mtbc committed:
avoid needless object creation

mtbc commented:
As noted on the card, this PR includes commits from https://github.com/openmicroscopy/openmicroscopy/pull/4957. This will cause intermittent "is not a member of group" errors in integration tests that do imports so do not worry about surprise failures in the test jobs unless they are consistent.

mtbc committed:
make session reload robust against user, group name change

cf. trac 11693; cc: @joshmoore

mtbc committed:
prefix light admin privilege config property names

mtbc committed:
adjust admin service test to expect admin privilege prefixes

mtbc committed:
do __del__() on created OMERO clients in integration test

mtbc committed:
added WriteManagedRepo, DeleteManagedRepo admin privileges

mtbc committed:
introduce and wire bean for noting managed repo UUIDs

mtbc committed:
actually note managed repo UUIDs

mtbc committed:
wire OMERO interceptor into Repository DAO

mtbc committed:
give OMERO interceptor another chance to check file creation

Needs another go because now repo is set.
This can be reverted once repo is no longer set via SQL.

Because OriginalFile is originally created with null repo,
light admins require WriteFile regardless of repository.

mtbc committed:
basic sec. system checks WriteManagedRepo, DeleteManagedRepo

mtbc committed:
adjust integ. tests for WriteManagedRepo, DeleteManagedRepo

mtbc committed:
reorder integration tests

mtbc committed:
set creation time for newly created files

mtbc committed:
fix indent to match surrounding

pwalczysko committed:
Merge branch 'roles' of github.com:mtbc/openmicroscopy into new-role

pwalczysko committed:
Fix testDataOrganizerChownAll test to pass after Marks additions to permission granularity

pwalczysko committed:
Temporary commit to delete or squash later

pwalczysko committed:
Fix testImporterAsNoSudoChownOnly and testImporterAsSudoChown for new Marks features

mtbc committed:
use correct group context for finalizing import log file

mtbc committed:
adjust linebreak in comment

mtbc committed:
graph ops must allow perms check for some object properties

pwalczysko committed:
Fix the comments in two tests

mtbc committed:
fix light admin script repo test to use correct repo

pwalczysko committed:
Merge branch 'roles' of github.com:mtbc/openmicroscopy into new-role

mtbc committed:
return graph ops to previous treatment of groups

partially reverts 229a7e01
also avoids loading OriginalFile instances

pwalczysko committed:
Add check for OrigFile to testImporterAsNoSudoChownOnlyWorkflow and fix testImporterAsSudoChown for new Marks permissions

mtbc committed:
do not require delete permission for Chown light admins

mtbc committed:
remove unused imports from graphs commands

pwalczysko committed:
Fix testImporterDelete for new permissions added by Mark

pwalczysko committed:
Merge branch 'roles' of github.com:mtbc/openmicroscopy into new-role

pwalczysko committed:
Adjust the testImporterAsSudoChown for the new behaviour of Chown (draws with it all the other necessary permissions)

pwalczysko committed:
Fix testImporterAsNoSudoChownOnlyWorkflow for new permissions added by mark with changed chown width

pwalczysko commented:
@mtbc : so now that Chown is dragging with itself some delete privileges, in order to be able to perform at all in most situation, do we have a definition about which ones precisely ? In other words, what additional (probably deletion) actions will a light admin equipped with Chown be able to do (without being explicitly given any additional privileges) ? Checked the Experimenter_config table in the DB, even when Chown is true, the Delete rows are still false, so it is not explicitly obvous what additional perms were given (maybe none, which would be greatttt....)

pwalczysko committed:
Improve comments on testImporterAsSudoChown

pwalczysko committed:
Fix testDataOrganizerChownAll for new chown behaviour and adjust comments

pwalczysko committed:
Fix testImporterDelete for new permissions, simplify

mtbc commented:
The extra delete power now coming with `Chown` should apply *only* when actually doing a chown, not in other circumstances, so other operations should remain as before.

pwalczysko commented:
yupeee. thanks @mtbc

pwalczysko committed:
Fix testImporterAsNoSudoChgrpChownWorkflow for new permissions

pwalczysko committed:
Adjust testImporterAsSudoChgrp for new permissions, simplify

pwalczysko committed:
Simplify testLightAdminEdit for new permissions

pwalczysko committed:
Simplify testImporterAsSudoCreateImport for new permissions setup

pwalczysko committed:
Simplify testImporterAsNoSudoLinkInTargetGroup for new permissions setup

pwalczysko committed:
Remove unused data providers

pwalczysko commented:
I have the tests running now with Mark's new permissions setup. Also simplified them as much as possbile in order to spare resources - more on this to come.
It seems we have 9 working workflows (back end tested) as outlined in the .ppt https://github.com/openmicroscopy/openmicroscopy/pull/5048#issuecomment-273841100.
Some are "slick", some more clanky. The most used most clanky workflow would be the linking (put someones image into his dataset and then change the ownership of the link) -> additionally this one does not work in private group at all as already discussed.

ToDo for me:
  - write test for User Organizer (user/group manipulator)
  - try to get more light admin workflows out of the nine above checked using clients
  - write test for WriteScriptRepo (as this is rather a safety feature, maybe I could fall back on a test which was already written by @mtbc ?)

pwalczysko commented:
The LightAdminRoles test takes now 25 minutes on my mac.

jburel commented:
@pwalczysko any way we can break it?

pwalczysko commented:
@jburel : We could split the client-side checking (mainly CLI workflows based on the 9 workflows in the test should be checked). For that, one needs to create the light admin and a user(s) according to the setup and necessary permissions bullet points on each slide in my ppt  and go through the workflow on the CLI. But the person testing this must be on this branch here of course.

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -0,0 +1,1330 @@
+           String groupPermissions) throws Exception {
+       /* only DeleteOwned permission is truly needed for deletion of links, dataset
+        * and image (with original file) when not sudoing */
+       boolean deletePassing = (!isSudoing && permDeleteOwned) || isSudoing;
```
this is simply,
```java
final boolean deletePassing = permDeleteOwned || isSudoing;
```

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -0,0 +1,1330 @@
+               doChange(Requests.delete().target(sentDat).build());
+               doChange(Requests.delete().target(sentProj).build());
+           } catch (ServerError se) {
+           /* not expected */
```
If it's not expected then why catch it?
Also, if we must catch it, maybe at least have an `Assert.fail` here then.

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -0,0 +1,1330 @@
+     */
+
+    @Test(dataProvider = "6 privileges cases")
+    public void testImporterAsNoSudoChownOnlyWorkflow(boolean isAdmin, boolean permWriteOwned, boolean permWriteManagedRepo, boolean permWriteFile, boolean permChown,
```
`permWriteOwned` seems to be called `isSudoing` in the provider!

pwalczysko committed:
Remove extra permission given in testImporterAsNoSudoChgrpChownWorkflow

pwalczysko committed:
Rename all the tests to match the content, improve one comment

pwalczysko commented:
The Power Point presentation is now as a [g.doc](https://docs.google.com/presentation/d/18kc75LHiItkHr_dvG1R6jW1qmTNayan8x4ZQC6s7XmI/edit#slide=id.g1b11290c7c_2_96) better accessible.

pwalczysko committed:
Fix testDelete cf. Mark's comments

pwalczysko commented:
@mtbc : This should fix the issues you raised. The issue with badly named data provider variables I will deal with once we have all the tests.

pwalczysko committed:
Add "severing ot the link" workflow to testChgrp

pwalczysko committed:
Add "severing of the link" workflow to testChown

pwalczysko committed:
Improve testLinkNoSudo description

pwalczysko committed:
Add test for rendering settings creation and chown by light admin

pwalczysko commented:
@mtbc : Would you fly over https://github.com/openmicroscopy/openmicroscopy/pull/5048/commits/eb738dd0e3ec0d561c37081bb0c902940c8d119d briefly please ? I hope it is doing what I say in the comments. Definitely no regression/difference from the full admin behaviour on the Rnd settings front.

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -1249,6 +1253,59 @@ public void testChownAllBelongingToUser(boolean isAdmin, boolean permChgrp, bool
         Assert.assertEquals(retrievedProjectDatasetLinkOtherGroup2.getDetails().getOwner().getId().getValue(), recipient.userId);
     }
 
+    /** Test of light amdin without using Sudo.
```
admin :smiley:

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -1249,6 +1253,59 @@ public void testChownAllBelongingToUser(boolean isAdmin, boolean permChgrp, bool
+        boolean isExpectSuccessCreateAndChown = isExpectSuccessCreateRndSettings && permChown;
+        final EventContext normalUser = newUserAndGroup(groupPermissions);
+        /* set up the light admin's permissions for this test */
+        ArrayList <String> permissions = new ArrayList <String>();
```
```java
List<String> permissions = new ArrayList<String>();
```
(use most general class for variable declaration, don't precede angle brackets with spaces)

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -1249,6 +1253,59 @@ public void testChownAllBelongingToUser(boolean isAdmin, boolean permChgrp, bool
+        final EventContext normalUser = newUserAndGroup(groupPermissions);
+        /* set up the light admin's permissions for this test */
+        ArrayList <String> permissions = new ArrayList <String>();
+        if (permChown) permissions.add(AdminPrivilegeChown.value);;
```
double semicolon

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -1249,6 +1253,59 @@ public void testChownAllBelongingToUser(boolean isAdmin, boolean permChgrp, bool
+        if (permChown) permissions.add(AdminPrivilegeChown.value);;
+        if (permWriteOwned) permissions.add(AdminPrivilegeWriteOwned.value);
+        final EventContext lightAdmin;
+        lightAdmin = loginNewAdmin(isAdmin, permissions);
```
could do this after the image is created to avoid switching back between normal user and admin so much

mtbc commented on components/tools/OmeroJava/test/integration/LightAdminRolesTest.java:
```diff
@@ -1249,6 +1253,59 @@ public void testChownAllBelongingToUser(boolean isAdmin, boolean permChgrp, bool
+        lightAdmin = loginNewAdmin(isAdmin, permissions);
+        /* create an image with pixels as normalUser in a group of the normalUser */
+        loginUser(normalUser);
+        client.getImplicitContext().put("omero.group", Long.toString(normalUser.groupId));
```
the normal user should already be in their group

pwalczysko committed:
Improve logic of testRenderingSettingsNoSudo

pwalczysko committed:
Fix typo

pwalczysko committed:
Add test for ROI creation and chowning

pwalczysko commented:
The last commit https://github.com/openmicroscopy/openmicroscopy/pull/5048/commits/3d02afb9150473089791990591528f799702b9af adds test for ROIs. We have no difference between root and light admin behaviour there (no regression). One inexpected behaviour detected, but not urgent/blocking, basically chown of ROIs in read-only is not possible per se if the image is not included, see [trello card](https://trello.com/c/7o4q2Tkt/155-fix-graphs-for-mixed-ownership-read-only)

pwalczysko committed:
Replace the declaration in all tests cf. Mark's comment

pwalczysko committed:
Fix double semicolon typo

pwalczysko committed:
Swap the order of image creation and light admin establishment cf. Mark's comment

pwalczysko committed:
Remove the group context line as the user is already in the group

pwalczysko committed:
Add test for File Attachment creation and linking

pwalczysko committed:
Rename the testFileAttachmentNoSudo (no tags are being handled in this test)

pwalczysko committed:
Add test for upload of official script

pwalczysko committed:
Add test for Deleting of official script

pwalczysko committed:
Add test for adding user into group via iAdmin.addGroups

pwalczysko committed:
Add test for removal of a user from a group

pwalczysko committed:
Add test for setting a new group owner

pwalczysko committed:
Add test for unsetting Group Owner

will-moore commented:
I think it's currently possible for 'root' to add restrictions to themselves which we should probably disallow at the server level (in the same way that we currently prevent 'root' from removing themselves from 'system' group).

mtbc commented:
Re. https://github.com/openmicroscopy/openmicroscopy/pull/5048#issuecomment-278060296 have tried requesting access a couple of times but with no fruit.

will-moore commented:
It seems that a 'restricted admin' with 'ModifyUser' privilege can edit themselves OR another user to make them a full Admin!

mtbc commented:
Yes, "fix privilege elevation routes" is the current checklist item on https://trello.com/c/t0nT7KYa/133-new-role. :smiley:

pwalczysko commented:
@mtbc : Re https://github.com/openmicroscopy/openmicroscopy/pull/5048#issuecomment-279688557 - sorry, this shoiuld be now fixed, and the .ppt is accessible (albeit it is another .ppt, so please use the link as given in the header of this PR)

pwalczysko committed:
Add test for creation of new user

pwalczysko committed:
Add test for editing a user

pwalczysko committed:
Add test for group creation

pwalczysko committed:
Correct the privilege line in three tests

pwalczysko committed:
Add test for group editing

pwalczysko commented:
Testing performance:

On new-role branch:
 - on clean DB locally with freshly restarted laptop running ``.../integration/chown/PermissionsTest.java`` takes 284588 ms = 4.7 minutes (this test has 103 methods, resulting in at least 103 groups and users)
  - importing a plate as root into System  ``bin/omero import /Volumes/ome/data_repo/curated/cellworx/pki/`` results in ``193 files uploaded, 1 fileset, 1 plate created, 384 images imported, 0 errors in 0:02:51.375``
  - moving a plate as root (owner) to a different, newly created group called test (root is a member of this group too)
```
[pwalczysko@ls31619 ~/Work/openmicroscopy/dist]$ time bin/omero chgrp test Plate:10
Using session 5b83ee04-81a6-4bdf-b543-b385e5491614 (root@localhost:4064). Idle timeout: 10 min. Current group: system
omero.cmd.Chgrp2 Plate:10 ok

real	0m5.742s
user	0m0.526s
sys	0m0.117s
```
  - deleting the Plate
```
time bin/omero delete Plate:10
Using session 5b83ee04-81a6-4bdf-b543-b385e5491614 (root@localhost:4064). Idle timeout: 10 min. Current group: system
omero.cmd.Delete2 Plate:10 ok

real	0m3.136s
user	0m0.516s
sys	0m0.120s
```

On develop branch (``latest``)
 - on clean DB locally with freshly restarted laptop running ``.../integration/chown/PermissionsTest.java`` takes 222193 ms = 3.7 minutes (this test has 103 methods, resulting in at least 103 groups and users)
 - mporting a plate as root into System  ``bin/omero import /Volumes/ome/data_repo/curated/cellworx/pki/`` results in``193 files uploaded, 1 fileset, 1 plate created, 384 images imported, 0 errors in 0:01:56.001
``
 - moving the plate into a new group test
```
time bin/omero chgrp test Plate:10
Using session 2867b750-0c6a-4e28-b192-c0baf8fdd343 (root@localhost:4064). Idle timeout: 10 min. Current group: system
omero.cmd.Chgrp2 Plate:10 ok

real	0m5.896s
user	0m0.544s
sys	0m0.122s
```
 - deleting the plate
```
time bin/omero delete Plate:10
Using session 2867b750-0c6a-4e28-b192-c0baf8fdd343 (root@localhost:4064). Idle timeout: 10 min. Current group: system
omero.cmd.Delete2 Plate:10 ok

real	0m3.130s
user	0m0.521s
sys	0m0.116s
```

pwalczysko commented:
@mtbc : Looks like no difference on move and delete of a real plate with 384 images. Some 28-47 percent slower on import and running integration tests (chown).

Will repeat the import of the plate in the morning maybe there is varying performance of my Mac.

Edit: the morning result on roles branch for import gave
``==> Summary
193 files uploaded, 1 fileset, 1 plate created, 384 images imported, 0 errors in 0:02:41.298``
The variation seems to be some 10 secs,  the worsening 38 percent, I think the 30-47 percent above still stands.

pwalczysko committed:
Improve javadoc, spacing and remove unused imports

pwalczysko committed:
Add more precise provider for testCreateLinkImport and testEdit

pwalczysko committed:
Add more precise provider for testDelete, improve comments in the test

pwalczysko committed:
Add more precise provider for testChgrp, improve comments in the test

pwalczysko committed:
Add more precise provider for testChown, improve comments in the test

pwalczysko committed:
Add more precise provider for testImporterAsNoSudoChownOnlyWorkflow, improve comments in the test

pwalczysko committed:
Add more precise provider for testLinkNoSudo, improve test comments and structure

pwalczysko committed:
Add more precise provider for testROIAndRenderingSettingsNoSudo, improve comments in the test

pwalczysko committed:
Add more precise provider for testImporterAsNoSudoChgrpChownWorkflow

pwalczysko committed:
Add more precise provider to testChownAllBelongingToUser, remove redundant parameter from the test

pwalczysko committed:
Add more precise provider to all the rest of the tests

pwalczysko commented:
My 21 tests now take some 22 minutes. This makes it some ~1 minute per test. In the next step, I intend to cut the time to a half approx, as I will remove all isAdmin loops -> this is being tested already by @mtbc 's tests.

mtbc commented:
Note that my tests usually assume that the light admin has many privileges: I am focused on security testing that a restriction is disabling. Your tests often come from the other side, that a workflow is possible given only certain privileges.

pwalczysko commented:
Discussed with @mtbc : It is fine to remove the isAdmin = false cases, which is what I am talking about in https://github.com/openmicroscopy/openmicroscopy/pull/5048#issuecomment-282695999 - these cases are actually testing the isRestricted cases in Mark's tests again.

pwalczysko committed:
Rename the variables in simpler tests to isPrivileged to match the logic in Mark's tests

pwalczysko committed:
Remove unused data providers, improve comments on the other providers

pwalczysko committed:
Remove isAdmin from the tests

pwalczysko commented:
This PR was superseded by https://github.com/openmicroscopy/openmicroscopy/pull/5159. Closing.
